main/ZgatewayBT.ino

```diff
@@ -581,73 +583,78 @@ bool BTtoMQTT() {
-          for (int i = 0; i < 6; i++) {
-            extract_char(token_char, d[i].extract, d[i].start, d[i].len, d[i].reverse, false);
-            if (i == 3)
-              d[5].len = (int)strtol(d[i].extract, NULL, 16) * 2; // extracting the length of the rest data
-          }
+      String token = returnedString.substring(pos + BLEdelimiterLength, eolPos); //capture a BT device frame
+      returnedString.remove(0, eolPos + CRLR_Length); //remove frame from main buffer (including 'CR LF' chars)
+      Log.trace(F("Token: %s" CR), token.c_str());
+      if (token.length() > 32) { // we extract data only if we have size is at least the size of (MAC, TYPE, RSSI, and Rest Data Length)
+        String mac = F("");
+        mac.reserve(17);
+        for (int i = d[0].start; i < (d[0].start + d[0].len); i += 2) {
+          mac += token.substring((d[0].start + d[0].len) - i - 2, (d[0].start + d[0].len) - i);
+          if (i < (d[0].start + d[0].len) - 2)
+            mac += ":";
+        }
+        mac.toUpperCase();
 
-          if ((strlen(d[0].extract)) == 12) // if a BLE device is detected we analyse it
-          {
-            Log.trace(F("Creating BLE buffer" CR));
-            StaticJsonBuffer<JSON_MSG_BUFFER> jsonBuffer;
-            JsonObject& BLEdata = jsonBuffer.createObject();
-            strupp(d[0].extract);
-            String Id;
-            for (int i = 0; i < 12; i = i + 2) {
-              Id += String(d[0].extract[i]);
-              Id += String(d[0].extract[i + 1]);
-              if (i != 10)
-                Id += ":";
-            }
-            Log.trace(F("Id %s" CR), (char*)Id.c_str());
-            BLEdata.set("id", (char*)Id.c_str());
-            BLEdevice* device = getDeviceByMac((char*)Id.c_str());
-
-            if (isBlack(device))
-              return false; //if black listed mac we go out
-            if (oneWhite && !isWhite(device))
-              return false; //if we have at least one white mac and this mac is not white we go out
-            int rssi = (int)strtol(d[2].extract, NULL, 16) - 256;
-            BLEdata.set("rssi", (int)rssi);
+        String rssiStr = token.substring(d[2].start, (d[2].start + d[2].len));
+        int rssi = (int)strtol(rssiStr.c_str(), NULL, 16) - 256;
+
+        String restDataLengthStr = token.substring(d[3].start, (d[3].start + d[3].len));
+        int restDataLength = (int)strtol(restDataLengthStr.c_str(), NULL, 16) * 2;
+
+        String restData = F("");
+        if (restDataLength <= 60)
+          restData = token.substring(d[5].start, (d[5].start + restDataLength));
+
+        Log.trace(F("Creating BLE buffer" CR));
+        StaticJsonBuffer<JSON_MSG_BUFFER> jsonBuffer;
+        JsonObject& BLEdata = jsonBuffer.createObject();
+
+        Log.trace(F("Id %s" CR), (char*)mac.c_str());
+        BLEdata.set("id", (char*)mac.c_str());
+        BLEdevice* device = getDeviceByMac((char*)mac.c_str());
```

Comment:
You will need to merge the dev branch before an integration, especialy the modifications done [here ](https://github.com/1technophile/OpenMQTTGateway/pull/720)and [here](https://github.com/1technophile/OpenMQTTGateway/commit/0cd2464bf11c61b37f6e70fed1d86bac3f5860c4)

Reply:
Both of the commits mentioned are in this [branch commit history](https://github.com/jmw6773/OpenMQTTGateway/commits/HM10---BT-parsing). 

I did a quick glance over the code and the changes appear to be this this PR code. I can try re-basing to the `dev` branch again if you think I'm missing something. (Still new to rebase, so it's highly possible.)

Regarding the modifications mentioned, I'm pulling the MAC address directly from the BLE `token` string instead of using the `extract_char` function. This elements the `for` loop that was previously added, and uses the `start` and `len` values defined in the decompose struct. It should be injecting `:` into the correct places. (Couldn't get the relevant code block to inline.)
https://github.com/1technophile/OpenMQTTGateway/pull/724/files#diff-f711e1aab0b74a5422256770df31e7a1R617-R624

Reply:
You don't need to rebase, my mistake.